pkg/config/expand.go

```diff
@@ -166,6 +167,44 @@ func (bp Blueprint) expandBackend(grp *Group) {
 	}
 }
 
+func (bp Blueprint) expandProviders(grp *Group) {
+	// 1. DEFAULT: use TerraformProviders provider dictionary (if supplied)
+	// 2. If top-level TerraformProviders is defined, insert that
+	//    provider dictionary into resource groups which have no explicit
+	//    TerraformProviders
+	defaults := bp.TerraformProviders
+	pv := &grp.TerraformProviders
+	if len(defaults) == 0 && len(grp.TerraformProviders) == 0 {
+		gglConf := Dict{}
+		for s, v := range map[string]string{
+			"project": "project_id",
+			"region":  "region",
+			"zone":    "zone"} {
+			if bp.Vars.Has(v) {
+				gglConf = gglConf.With(s, GlobalRef(v).AsValue())
+			}
+		}
+		(*pv) = map[string]TerraformProvider{
+			"google": TerraformProvider{
+				Source:        "hashicorp/google",
+				Version:       ">= 4.84.0, < 5.30.0",
+				Configuration: gglConf},
+			"google-beta": TerraformProvider{
+				Source:        "hashicorp/google-beta",
+				Version:       ">= 4.84.0, < 5.30.0",
+				Configuration: gglConf}}
+	} else {
```

Comment:
If I'm reading this right, there is a bug, 

```go
	{ // no def PR, no group PR - match default values
		g := Group{Name: "clown"}
		noDefPr.expandProviders(&g)
		c.Check(g.TerraformProviders, DeepEquals, map[string]PR{
			"google": TerraformProvider{
				Source:  "hashicorp/google",
				Version: ">= 4.84.0, < 5.30.0"},
			"google-beta": TerraformProvider{
				Source:  "hashicorp/google-beta",
				Version: ">= 4.84.0, < 5.30.0"}})
	}
```

In that case we probably want configuration set for default google providers.


I would recommend you to:
* get rid of this large `if-else` statement
* have a separate function `getDefaultGoogleProviders`;
* have `defaults := bp.Providers or getDefaultGoogleProviders(bp)`

I'm on fence about **always** adding default google providers.

Reply:
In the test snipped above, shouldn't "google" and "google-beta" providers have "Configuration" set as well ?

Reply:
No, because there are no `universe_domain` or `custom_<service>_endpoint` settings for the default.  Configuration only gets populated when we're moving away from the default.